### PR TITLE
txngroup-deltas: Fix pointer bug copying deltas for txngroups

### PR DESCRIPTION
**Pull request by Eric-Warehime**
## Summary

The KvMods (and other fields in the StateDelta which are maps) have their values deleted when the roundCoW gets recycled. Because we're not copying this data out of the delta in the txn tracer, this affects the results returned by the tracer.

I've now copied all of the fields out of the maps in the StateDelta (for those that we care about returning).

## Test Plan

Adding full unit tests for kvmods--other data should function similarly. This required adding some functionality to the eval test ledger which was not initially implemented during the Boxes PR.

### CODE REVIEW

File: ledger/eval/txntracer.go
Comment (algochoi) [inline ledger/eval/txntracer.go:51]: I'm still not sure why we have to copy over the maps here (particularly what recycling the roundCow means)

Maybe an inline comment here will help explain the deep copying too

**Review by algochoi — COMMENTED** [review]
Nice catch, adding the box in the txntracer_test seems to show that the map is being overriden. I had a question in PR comment below

File: ledger/eval/txntracer.go
Comment (tzaffi) [inline ledger/eval/txntracer.go:61]: Codecov is complaining: Added line #L61 was not covered by tests

Is this possibly covered in a different package?

File: ledger/eval/txntracer.go
Comment (Eric-Warehime) [reply ledger/eval/txntracer.go:51]: I'll try to add a comment w/ some context. Here's the long version:

The `StateDelta` that we get for each transaction group is stored in a `roundCowState` [here](https://github.com/algorand/go-algorand/blob/master/ledger/eval/cow.go#L75). We create a separate `cow` for each transaction group ([eval.state::child](https://github.com/algorand/go-algorand/blob/master/ledger/eval/eval.go#L958)) and when we finish evaluating the group, the deltas get merged into the parent and the `cow` gets reused for the next group with a call to [cow::recycle](https://github.com/algorand/go-algorand/blob/master/ledger/eval/eval.go#L959). 

The `recycle` function goes through and [resets](https://github.com/algorand/go-algorand/blob/master/ledger/eval/cow.go#L337) the delta which deletes the keys in all of the maps.

Even though the [deltas](https://github.com/algorand/go-algorand/blob/master/ledger/eval/cow.go#L126) function returns by value, the maps are all pointers, so the value returned has pointers to the same objects which will eventually have all of their entries deleted as evaluation progresses.

File: ledger/eval/txntracer.go
Comment (Eric-Warehime) [reply ledger/eval/txntracer.go:61]: No I don't think it's covered, I'll add a lease to the txns in the tracer test to make sure it's covered.

Comment (tzaffi) [reply ledger/eval/txntracer.go:61]: Btw, another example that gets me pining for go 1.18+ generics. We could have one liner copies.

**Review by tzaffi — COMMENTED** [review]
Looks good. One thought crosses my mind. The current box app creates the box. I'm wondering if there could be a benefit to testing some of the following - or these may be overkill:
* modify a box with `box_replace`
* delete a box with `box_del`
* compose a `box_create` with a `box_del` in the same app call and/or in subsequent calls of the same group, or subsequent inner app calls.

**Comment by tzaffi** [comment]
> 
> I can add a bit more, but in general none of this code cares what the TEAL code is doing, just that it properly conveys the deltas associated with it. And the fact that KvMods are present pretty much means all KvMod opcode deltas will be present as well.

Totally up to you. If you're viewing the KvMods as a "black box" in which case the current scenario proves that the PR reveals the black box _as is_, then no further tests are required. But if the logic could be tricked into believing that there aren't any txndelta's when actually there were (e.g. by piling on changes, or reversing them) then another test case would be called for.

> 
> The inner txn box mods would probably be worthwhile to test. Potentially having subsequent updates to the same box as well I guess?

Yeah, in the past, the inner txns seemed to always be the most error prone due to recursion.

File: ledger/eval/txntracer.go
Comment (algochoi) [reply ledger/eval/txntracer.go:54]: appreciate the detailed context - that's why we are deep copying them right after transaction group has been evaluated but before the round has passed and state has been recycled.

**Review by algochoi — DISMISSED** [review]
LGTM, thanks for addressing comments!

**Comment by Eric-Warehime** [comment]
@tzaffi Yep, for the most part I'd like to treat it as an opaque operation, but that operation was only intended to be correct at the block level and not necessarily at the txn-group level so it's worthwhile to add more tests. I've added a `box_del` call in an inner txn to the test.

Let me know if you feel like we really need to test anything else that's critical path w/r/t the deltas here.

File: ledger/eval/txntracer_test.go
Comment (tzaffi) [inline ledger/eval/txntracer_test.go:324]: So this ghost of boxes past is picked up because the box was created then deleted?

**Review by tzaffi — APPROVED** [review]
LGTM

**Review by algochoi — DISMISSED** [review]

File: ledger/eval/txntracer_test.go
Comment (jannotti) [inline ledger/eval/txntracer_test.go:275]: I guess you didn't put these here originally, but the fact that we're editing hex bytecode here seems bad.  I assume this is a compiled program, can we use `logic.AssembleString` so it makes more sense?

File: ledger/eval/eval_test.go
Comment (jannotti) [inline ledger/eval/eval_test.go:603]: Every test that uses this `evalTestLedger` creeps me out.  It is a mock that maybe has the same behaviour as a real ledger, and maybe not. So every time we extend it to try to do more, it feels even more as though every test should be converted to use a real ledger.  `DoubleLedger` and `TestConsensusRange` have some real benefits in terms of accuracy and testing across changes in old versions.

I'm not going to hold up this PR, but we have to wean ourselves off this thing.

File: ledger/eval/eval_test.go
Comment (jannotti) [inline ledger/eval/eval_test.go:735]: This is the kind of craziness I'm talking about.  Shouldn't we use a real ledger, and no constantly make little updates to approximate it?

**Review by jannotti — APPROVED** [review]
I'm willing to merge this, let me know if you want to do anything about the things I mentioned.

@jasonpaulos should also see it though - it feels like you're both addressing some copying issues right now.

File: ledger/eval/txntracer.go
Comment (jasonpaulos) [inline ledger/eval/txntracer.go:71]: `delta.Accts` is also subject to recycling. I suspect you're not seeing a difference in your test because the `reset` function is not immediately destructive, but the underlying memory is still reserved for the next user of the cow.

https://github.com/algorand/go-algorand/blob/b988971925f5c9e7399c0df0393bb99f7376b1a6/ledger/ledgercore/statedelta.go#L258-L260

File: ledger/eval/txntracer_test.go
Comment (Eric-Warehime) [reply ledger/eval/txntracer_test.go]: Updated. I actually did write those originally, just not in this PR. I wasn't aware that this was the proper way to encode these.

File: ledger/eval/eval_test.go
Comment (Eric-Warehime) [reply ledger/eval/eval_test.go:603]: Yeah I agree the actual implementation of the test ledger is pretty far from the actual ledger. It does seem difficult to mock since the data retrieved in the actual ledger hits the caches (trackers) first and then disk if the cache misses.

I haven't looked into the `DoubleLedger` or `TestConsensusRange`, but I will do next time I need to test things requiring the ledger.

File: ledger/eval/txntracer.go
Comment (Eric-Warehime) [reply ledger/eval/txntracer.go]: This is tricky. It seems like you're right so I've copied all of the data out of the AccountDeltas as well, but it is hard for me to reason about which CoWs are being reused across evaluation of multiple txn groups. 

I even added a second group in the same block and it didn't affect the returned results.

Comment (jasonpaulos) [reply ledger/eval/txntracer.go]: Before your fix in bdf35d8, when I run this test with `-count=100`, I see some errors, so it's definitely an issue.

Strangely adding the `-race` flag doesn't seem to catch this, though it has caught similar mistakes I've made around cow recycling.

The fix looks good and I don't see any more errors when testing locally.

File: ledger/eval/txntracer.go
Comment (jasonpaulos) [inline ledger/eval/txntracer.go:54]: nit: I'd recommend instantiating the maps with the right capacity, since you know what it is.

**Review by jasonpaulos — APPROVED** [review]

**Review by algorandskiy — APPROVED** [review]

**Review by algorandskiy — APPROVED** [review]

File: ledger/eval/txntracer.go
Comment (algorandskiy) [reply ledger/eval/txntracer.go:54]: +1 to Jason

**Review by jasonpaulos — APPROVED** [review]

**Review by algorandskiy — APPROVED** [review]

**Review by algorandskiy — APPROVED** [review]

**Pull request merged**
